kernel/async/monitor-common/async/TaskletQueue.hh

```diff
@@ -28,108 +28,155 @@
-
-  std::atomic<TaskletBase*> privateTail;
-  //char padding[64]; // to ensure separate cache lines
-  std::atomic<TaskletBase*> sharedTail; // TODO should be on a separate local cacheline
-};
-
-inline bool TaskletQueue::tryAcquire() {
-  TaskletBase* oldtail = FREE();
-  return sharedTail.compare_exchange_strong(oldtail, LOCKED(), std::memory_order_relaxed);
-}
-  
-inline bool TaskletQueue::pushPrivate(Tasklet* t) {
-  TaskletBase* oldtail = privateTail.exchange(t, std::memory_order_relaxed);
-  t->nextTasklet.store(oldtail, std::memory_order_relaxed);
-  return oldtail == FREE(); // true if this was the first message in the queue
-}
-
-inline bool TaskletQueue::push(Tasklet* t) {
-  t->nextTasklet.store(INCOMPLETE(), std::memory_order_relaxed); // mark as incomplete push
-  TaskletBase* oldtail = sharedTail.exchange(t, std::memory_order_release); // replace the tail
-  t->nextTasklet.store(oldtail, std::memory_order_relaxed); // comlete the push
-  return oldtail == FREE(); // true if this was the first message in the queue
-}
-
-inline Tasklet* TaskletQueue::pull() {
-  // try to retrieve from the private queueu 
-  TaskletBase* oldtail = privateTail.exchange(FREE(), std::memory_order_relaxed); // avoid load()
-  ASSERT(oldtail != LOCKED());
-  if (oldtail != FREE()) {
-    privateTail.store(oldtail->nextTasklet.load(std::memory_order_relaxed),
-		      std::memory_order_relaxed); // remove oldtail from private queue
-    return static_cast<Tasklet*>(oldtail);
-  } // else try to retrieve from the shared queue
-  oldtail = sharedTail.exchange(LOCKED(), std::memory_order_relaxed); // detach tail and mark as locked
-  if (oldtail == FREE() || oldtail == LOCKED()) return nullptr; // nothing was in the shared queue, but now it is locked
-  while (!(oldtail == FREE() || oldtail == LOCKED())) {
-    TaskletBase* next;
-    do { // wait until the push is no longer marked as incomplete
-      // use exchange in order to avoid shared state of cacheline
-      next = oldtail->nextTasklet.exchange(INCOMPLETE(), std::memory_order_acquire);
-      if (next == INCOMPLETE()) hwthread_pause(); // sleep for a short amount of cycles
-    } while (next == INCOMPLETE());
-    if (next == FREE() || next == LOCKED()) {
-      return static_cast<Tasklet*>(oldtail); // directly return last task from old shared queue
-    } // else push it into the private queue
-    oldtail->nextTasklet.store(privateTail.exchange(oldtail, std::memory_order_relaxed),
-			       std::memory_order_relaxed);
-    oldtail = next;
-  }
-  PANIC_MSG(false, "should never reach here");
-}
-
-inline bool TaskletQueue::tryRelease()
-{
-  ASSERT(privateTail == FREE());
-  TaskletBase* oldtail = LOCKED();
-  return sharedTail.compare_exchange_strong(oldtail, FREE(), std::memory_order_relaxed);
-}
-
-} // async
-} // mythos
+
+  class TaskletQueueBaseDefault {
+  public:
+    TaskletQueueBaseDefault()
+      :privateTail_(Tasklet::FREE), sharedTail_(Tasklet::FREE)
```

Comment:
I think, moving the initial value to the declaration is more compact and simpler:
`std::atomic<uintptr_t> privateTail_ = {Tasklet::FREE};`